src/mono/mono/mini/aot-compiler.c

```diff
@@ -6589,6 +6589,7 @@ emit_and_reloc_code (MonoAotCompile *acfg, MonoMethod *method, guint8 *code, gui
 						direct_call_target = symbol;
 						patch_info->type = MONO_PATCH_INFO_NONE;
 					} else if ((m_class_get_image (patch_info->data.method->klass) == acfg->image) && !got_only && is_direct_callable (acfg, method, patch_info)) {
+#if !defined(TARGET_IOS) || defined(TARGET_TVOS)
```

Comment:
why is this codepath skipped for iOS?

Reply:
Direct call transformation is causing issues with an empty dedup assembly in llvm debug/release configurations. By skipping it for iOS, calls are forced to retrieve offsets by using PLT and GOT entries.

For example, in HelloiOS app when using debug configuration, `NativeRuntimeEventSource` constructor throws when initializing its attribute `EventSource`, as `EventListener` is `NULL`.
https://github.com/dotnet/runtime/blob/8b2a92019b2bd24a7c069ef8dfc7d99c28f30d4b/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1566

Similarly, when using release configuration, `InternalWait` throws as `default` is `NULL`.
https://github.com/dotnet/runtime/blob/8b2a92019b2bd24a7c069ef8dfc7d99c28f30d4b/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs#L140

In both cases, it fails when using default or global variables.

Reply:
Thanks for the explanation. 
I am just wondering whether we lose the part of the optimization by excluding it for a specific platform, compared to adding more restrictions programatically in the conditions above.

Reply:
Direct call transformation hasn't been executed for iOS as `acfg->aot_opts.dedup` was `FALSE`, so it shouldn't affect the existing optimizations. 

I agree with you that adding more restrictions programatically in the conditions would be a better option. With the dedup enabled, it fails when callee type has static fields. Expression like `cmethod->klass->has_static_fields == 0` might replace the existing directive, but it would affect web assembly optimizations.

I see the following options:
 - Use directive without affected optimizations
 - Use condition and directive with optimizations enabled for iOS

Reply:
It seems to be a [separate issue](https://github.com/dotnet/runtime/issues/82224) from the dedup. Additional size savings could be achieved. Not a blocker for the dedup.

Reply:
Thanks for checking this out!
Is it related to https://github.com/dotnet/runtime/pull/81319#discussion_r1107412751 ?

Reply:
Yes, I've tried the proposed approach but it seems that `is_direct_callable` method requires caller method (not callee) and a patch.